schnet/CFConv.h

```diff
@@ -0,0 +1,201 @@
+     */
+    virtual void build(const float* positions, const float* periodicBoxVectors) = 0;
+    /**
+     * Get whether the periodic box vectors specified in the most recent call to build() described
+     * a triclinic (not rectangular) box.
+     */
+    virtual bool getTriclinic() const = 0;
+    /**
+     * Get the number of atoms in the system.
+     */
+    int getNumAtoms() const {
+        return numAtoms;
+    }
+    /**
+     * Get the cutoff distance.
+     */
+    float getCutoff() const {
+        return cutoff;
+    }
+    /**
+     * Get whether to apply periodic boundary conditions.
+     */
+    bool getPeriodic() const {
+        return periodic;
+    }
+private:
+    int numAtoms;
+    float cutoff;
+    bool periodic;
+};
+
+/**
+ * This class computes the continuous filter convolution (cfconv) function used in SchNet.
+ * Create an instance of this class at the same time you create the model and then reuse it
+ * for every calculation on that model.
+ * 
+ * For each pair of atoms, it performs the following calculations:
+ * 
+ * 1. Compute a set of Gaussian basis functions describing the distance between them.
+ * 2. Pass them through a dense layer.
+ * 3. Apply a shifted softplus activation function.
+ * 4. Pass the result through a second dense layer.
+ * 5. Apply a cosine cutoff function to make interactions smoothly go to zero at the cutoff.
+ * 
+ * For each atom, the output is the sum over all neighbors of the above calculation multiplied
+ * by the neighbor's input vector.
+ * 
+ * This calculation is designed to match the behavior of SchNetPack.  It is similar but not
+ * identical to that described in the original SchNet publication.
+ * 
+ * This is an abstract class.  Subclasses provide implementations on particular
+ * types of hardware.
+ */
+class CFConv {
+public:
+    /**
+     * Construct on object for computing continuous filter convolution (cfconv) functions.
+     *
+     * @param numAtoms       the number of atoms in the system
+     * @param width          the number of elements in the input and output vectors
+     * @param numGaussians   the number of Gaussian basis functions to use, uniformly spaced between 0 and cutoff
+     * @param cutoff         the cutoff distance
+     * @param periodic       whether to apply periodic boundary conditions
+     * @param gaussianWidth  the width of the Gaussian basis functions
+     * @param w1             an array of shape [numGaussians][width] containing the weights of the first dense layer
+     * @param b1             an array of length [width] containing the biases of the first dense layer
+     * @param w2             an array of shape [width][width] containing the weights of the second dense layer
+     * @param b2             an array of length [width] containing the biases of the second dense layer
+     */
+    CFConv(int numAtoms, int width, int numGaussians, float cutoff, bool periodic, float gaussianWidth) :
```

Comment:
`w1`, `b1`,  etc. is described above, but they are missing here.

Reply:
Good catch, I should remove those.  The subclass constructors take those arguments, but they don't pass them on to the parent class.